File: 2tdsr/enterprise-application-development/01-revisao/src/br/com/fiap/exercicio/bean/Conta.java
package br.com.fiap.exercicio.bean;

import java.util.Calendar;

import br.com.fiap.exercicio.exception.SaldoInsuficienteException;

// Classe abstrata -> não pode ser instanciada e PODE conter métodos abstratos

public abstract class Conta {

	private Integer agencia;
	private Integer numero;
	private Calendar dataAbertura;
	protected Double saldo;

	public Conta(Integer agencia, Integer numero, Calendar dataAbertura, Double saldo) {
		this.agencia = agencia;
		this.numero = numero;
		this.dataAbertura = dataAbertura;
		this.saldo = saldo;
	}

	public Conta() {}

	public void depositar(Double valor) {
		saldo += valor;
	}

	public abstract void retirar(Double valor) throws SaldoInsuficienteException;

	// Getters e setters
	public Integer getAgencia() {
		return agencia;
	}

	public void setAgencia(Integer agencia) {
		this.agencia = agencia;
	}

	public Integer getNumero() {
		return numero;
	}

	public void setNumero(Integer numero) {
		this.numero = numero;
	}

	public Calendar getDataAbertura() {
		return dataAbertura;
	}

	public void setDataAbertura(Calendar dataAbertura) {
		this.dataAbertura = dataAbertura;
	}

	public Double getSaldo() {
		return saldo;
	}

	public void setSaldo(Double saldo) {
		this.saldo = saldo;
	}

}

File: 2tdsr/enterprise-application-development-asp-net/Fiap.Aula01/Fiap.Aula01.Exercicio01/Models/Conta.cs
using Fiap.Aula01.Exercicio01.Models;

namespace Fiap.Aula01.Exercicio01.Models
{
    public abstract class Conta
    {
        public int Agencia { get; set; }
        public int Numero { get; set; }
        public decimal Saldo { get; protected set; }
        public DateTime DataAbertura { get; set; }
        public IList<Cliente> Clientes { get; set; }

        public Conta(int agencia, int numero, IList<Cliente> clientes)
        {
            Agencia = agencia, Numero = numero, Clientes = clientes
        }

        public abstract void Depositar(decimal valor);

        public void Retirar(decimal valor) 
        {
            if(valor < 0) 
            {
                throw new ArgumentException("O valor deve ser maior do que zero")
            }
            Saldo += valor;
        }
    }
}